### ISSUE TITLE
authoring not working

### ISSUE THREAD

**Issue by samuelba** [google/copybara#83]
We want to use copybara with Gitlab. The workflow I imagine, since copybare doesn't support Gitlab as good as Github, is the following:

1. A user creates a pull request on the external repo.
2. We use copybara to copy (CHANGE_REQUEST) the branch of the pull request to our internal repo.
3. We do a pull request.
4. We merge this pull request.
5. We sync back (ITERATIVE) our internal source to the external repo.

However, I tried this and the only thing that gets displayed at the external repo is a commit of the "Merge branch 'external-001' into 'master'". There is no sign of the original author or of the original commit of the author.

I only see the commit of the original author in our internal repo.

My copy.bara.sky looks like this:
```
internalUrl = "git@gitlab.com:example/internal_test_repo.git"
externalUrl = "git@gitlab.com:example/external_test_repo.git"

core.workflow(
    name = "default",
    origin = git.origin(
        url = internalUrl,
        ref = "master",
    ),
    destination = git.destination(
        url = externalUrl,
        fetch = "master",
        push = "master",
    ),
    origin_files = glob(["project_001/share_001/**", "project_001/share_002/**", "project_002/subproject_002/**", "external/README.md"], exclude = ["**/key.txt"]),

    destination_files = glob(["**"], exclude = ["external_only.txt"]),

    mode = "ITERATIVE",
    authoring = authoring.pass_thru(default = "Example <example@gmail.com>"),
    transformations = [
        # Move to root of OSS repo.
        core.move("external/README.md", "README.md"),
        core.move("project_002", ""),
        metadata.restore_author(label='ORIGINAL_AUTHOR', search_all_changes=True),
    ],
)
core.workflow(
    name = "import_branch",
    origin = git.origin(
        url = externalUrl,
        ref = "external-001"
    ),
    destination = git.destination(
        url = internalUrl,
        fetch = "master",
        push = "external/external-001",
    ),
    origin_files = glob(["**"], exclude = ["bla.txt"]),

    destination_files = glob(["project_001/share_001/**", "project_001/share_002/**", "project_002/subproject_002/**", "external/README.md"], exclude = ["**/key.txt"]),

    mode = "CHANGE_REQUEST",
    authoring = authoring.pass_thru(default = "Example <example@gmail.com>"),
    transformations = [
        # Move from root of OSS repo.
        core.move("README.md", "external/README.md"),
        core.move("subproject_002", "project_002/subproject_002"),
        metadata.save_author(label='ORIGINAL_AUTHOR'),
    ],
)
```

**Comment by samuelba** [comment]
I think I found a solution, not sure though if this is the optimal/recommended solution. I use `first_parent = False,` in `git.origin()` and I use `ITERATIVE` mode for both processes. First for getting the external branch with all commits to the internal repository and than to sync back all commits from the internal repository to the external repository. This way I have not only the merge commits, but all commits, including the original author.

Maybe we could add an example like this to the examples? Of course only if this is the optimal/recommended solution for the task.

My copy.bara.sky looks like this now:
```
internalUrl = "git@gitlab.com:example/internal_test_repo.git"
externalUrl = "git@gitlab.com:example/external_test_repo.git"

core.workflow(
    name = "default",
    origin = git.origin(
        url = internalUrl,
        ref = "master",
        first_parent = False,
    ),
    destination = git.destination(
        url = externalUrl,
        fetch = "master",
        push = "master",
    ),
    origin_files = glob(["project_001/share_001/**", "project_001/share_002/**", "project_002/subproject_002/**", "external/README.md"], exclude = ["**/key.txt"]),

    destination_files = glob(["**"], exclude = ["external_only.txt"]),

    mode = "ITERATIVE",
    authoring = authoring.pass_thru(default = "Example <example@gmail.com>"),
    transformations = [
        # Move to root of OSS repo.
        core.move("external/README.md", "README.md"),
        core.move("project_002", ""),
        metadata.restore_author(label='ORIGINAL_AUTHOR', search_all_changes=True),
    ],
)
core.workflow(
    name = "import_branch",
    origin = git.origin(
        url = externalUrl,
        ref = "external-001",
        first_parent = False,
    ),
    destination = git.destination(
        url = internalUrl,
        fetch = "master",
        push = "external/external-001",
    ),
    origin_files = glob(["**"], exclude = ["bla.txt"]),

    destination_files = glob(["project_001/share_001/**", "project_001/share_002/**", "project_002/subproject_002/**", "external/README.md"], exclude = ["**/key.txt"]),

    mode = "ITERATIVE",
    authoring = authoring.pass_thru(default = "Example <example@gmail.com>"),
    transformations = [
        # Move from root of OSS repo.
        core.move("README.md", "external/README.md"),
        core.move("subproject_002", "project_002/subproject_002"),
        metadata.save_author(label='ORIGINAL_AUTHOR'),
    ],
)
```

**Comment by samuelba** [comment]
Ok, this is a very bad solution... it reverts changes of the SOT which where not yet synchronized with the external source.

**Comment by samuelba** [comment]
This means i stillt don't know how to import an external branch with all the commits into the internal source.

**Comment by mikelalcon** [comment]
Hi!

Sorry for the late response. If I understand correctly, the problem that you are facing is that Copybara internal -> external does a linear history export.

Copybara is designed to move linear history (no merges, etc.). But, we have a way to show the PR as 'merged'. The problem is that it only works for GitHub. But you could contribute one for Gitlab, should be easy (see GitHubPRIntegrateLabel and GerritIntegrateLabel as an example). We would also need a GitLab origin. This could be a wrapper of git.origin or github origin. We could skip this part if we added some sort of flag.

Assuming we had a GitLabPRIntegrateLabel, this would be used as:

add: metadata.expose_label("COPYBARA_INTEGRATE_REVIEW") as last transformation to pr import to internal. Yo should rebase/fast-forward the commit (no merge) so that the label is kept.

You end up internally with something like this:

* Internal change
* Change from a PR (has the label in body(
* Internal change

add: metadata.expose_label("COPYBARA_INTEGRATE_REVIEW") as last transformation from internal to external.

When GitDestination sees this, it runs the integrate label and does a fake or semifake merge. You end up externally with something like:

* Internal change
* Merge from PR xxxx
| \ 
|  * A change from the PR
|  * A change from the PR
|/
* Internal change

TL;DR; We would need a couple of things in Copybara to be implemented. Would you be willing to help ? :)

**Comment by samuelba** [comment]
Well the problem is that `CHANGE_REQUEST` only imports a "squashed" commit of the external branch. I would like to import all commits without squashing them, such that I can see all commits of the external branch in the internal branch.

However, I just testet `github_pr_origin` and `github_pr_destination` and it seems to do the same, combine all commits of the pull request into one commit. See:

Imported pull request has 1 commit: https://github.com/samuelba/internal_test_repo/pull/3
Original pull request has 3 commits: https://github.com/samuelba/external_test_repo/pull/3

So I guess implementing the Gitlab API doesn't help for that. But a Gitlab API to auto import pull requests and close pull requests would be still nice. I would be interested in contributing, but my Java knowledge is not that deep, so I would basically copy the Github files and try to adapt them for Gitlab including lots of try-and-error.

**Comment by mikelalcon** [comment]
I see. You could make it work with some scripting wrapping Copybara (that we could add native if this starts to be common pattern).

You import to a branch as you do now. Do the merge manually (but in the merge you add the integrate label or any text you want).

Then in the way out you run copybara and after that in the script you git log looking for that label and rewrite history to add a fake merge with the original PR. This would be ~40-50 lines of bash, no much.
